### DISCUSSION TITLE
How can I use the constraints API to determine if a specific column "Allows Duplicates"

### DISCUSSION THREAD

**Discussion by seancolsen** [mathesar-foundation/mathesar#769] [Developer Q&A] [answered]
I'm working to [Implement setting a column as non-nullable or unique](https://github.com/centerofci/mathesar/issues/499), and the first thing I'd like to do is _display_ to the user whether a column "Allows Duplicates", as the spec shows here:

![image](https://user-images.githubusercontent.com/42411/139328781-c7503d8e-d5b4-4311-b8d7-37c7049fe429.png)

When I make a request to `/tables/{id}/constraints/`, I get back:

```json
{
  "count": 2,
  "results": [
    {
      "id": 34,
      "name": "trees_latin_name_key",
      "type": "unique",
      "columns": [
        "latin_name"
      ]
    },
    {
      "id": 22,
      "name": "trees_pkey",
      "type": "primary",
      "columns": [
        "mathesar_id"
      ]
    }
  ]
}
```

From this response, I can tell that columns `latin_name` and `mathesar_id` do _not_ allow duplicates, and all other columns _do_ allow duplicates.

But what are the other possible responses I'd receive?

In the backend code [here](https://github.com/centerofci/mathesar/blob/d3b565ddaed0eb20dd11f013ab8202c16b91c36f/db/constraints/utils.py#L7), I see:

```py
class ConstraintType(Enum):
    FOREIGN_KEY = 'foreignkey'
    PRIMARY_KEY = 'primary'
    UNIQUE = 'unique'
    CHECK = 'check'
    EXCLUDE = 'exclude'
```

I'm assuming those are all the possible values I'd receive in the "type" field within the API response.

Here's the extent of my knowledge:

- 'foreignkey' -- not relevant
- 'primary' -- effectively same as 'unique' in this context
- 'unique' -- indicates that duplicates are not allowed
- 'check' -- not relevant
- 'exclude' -- ?????

**Is the `'exclude'` constraint type relevant here? I'm not familiar with that constraint type.**

@kgodey or @mathemancer

**Comment by kgodey** [comment, marked as answer]
Any columns with the `type` attribute set to `primary` or `unique` do not allow duplicates. Please note that if a column is a part of a multi-column constraint,  then individual columns should be set to "Allow Duplicates" since the individual column does allow duplicates.

Exclusion constraints are not relevant, see the bottom of https://www.postgresql.org/docs/9.4/ddl-constraints.html for more information about them. It's a PostgreSQL-specific constraint.
